integration-tests/src/tests/client/features/congestion_control.rs

```diff
@@ -0,0 +1,228 @@
+    }
+
+    // The immediate protocol upgrade needs to be set for this test to pass in
+    // the release branch where the protocol upgrade date is set.
+    std::env::set_var("NEAR_TESTS_IMMEDIATE_PROTOCOL_UPGRADE", "1");
+
+    let sender_id: AccountId = "test0".parse().unwrap();
+    let contract_id: AccountId = "contract.test0".parse().unwrap();
+    let mut env = setup_runtime(sender_id.clone());
+
+    let genesis_block = env.clients[0].chain.get_block_by_height(0).unwrap();
+    let signer = InMemorySigner::from_seed(sender_id.clone(), KeyType::ED25519, sender_id.as_str());
+    let mut nonce = 1;
+
+    // prepare a contract to call
+    let contract = near_test_contracts::rs_contract();
+    let create_contract_tx = SignedTransaction::create_contract(
+        nonce,
+        sender_id.clone(),
+        contract_id.clone(),
+        contract.to_vec(),
+        10 * 10u128.pow(24),
+        PublicKey::from_seed(KeyType::ED25519, contract_id.as_str()),
+        &signer,
+        *genesis_block.hash(),
+    );
+    // this adds the tx to the pool and then produces blocks until the tx result is available
+    env.execute_tx(create_contract_tx).unwrap().assert_success();
+    nonce += 1;
+
+    // Test the function call works as expected, ending in a gas exceeded error.
+    let block = env.clients[0].chain.get_head_block().unwrap();
+    let fn_tx = new_fn_call_100tgas(&mut nonce, &signer, contract_id.clone(), *block.hash());
+    let FinalExecutionStatus::Failure(TxExecutionError::ActionError(action_error)) =
+        env.execute_tx(fn_tx).unwrap().status
+    else {
+        panic!("test setup error: should result in action error")
+    };
+    assert_eq!(
+        action_error.kind,
+        ActionErrorKind::FunctionCallError(FunctionCallError::ExecutionError(
+            "Exceeded the prepaid gas.".to_owned()
+        )),
+        "test setup error: should result in gas exceeded error"
+    );
+
+    // Now, congest the network with ~1000 Pgas, enough to have some left after the protocol upgrade.
+    let block = env.clients[0].chain.get_head_block().unwrap();
+    for _ in 0..10000 {
+        let fn_tx = new_fn_call_100tgas(&mut nonce, &signer, contract_id.clone(), *block.hash());
+        // this only adds the tx to the pool, no chain progress is made
+        let response = env.clients[0].process_tx(fn_tx, false, false);
+        assert_eq!(response, ProcessTxResponse::ValidTx);
+    }
+
+    // Allow transactions to enter the chain
+    let tip = env.clients[0].chain.head().unwrap();
+    for i in 1..6 {
+        env.produce_block(0, tip.height + i);
+    }
+
+    // Ensure we are still in the old version and no congestion info is shared.
+    check_old_protocol(&env);
+
+    env.upgrade_protocol_to_latest_version();
+
+    // check we are in the new version
+    let block = env.clients[0].chain.get_head_block().unwrap();
+    assert!(ProtocolFeature::CongestionControl.enabled(block.header().latest_protocol_version()));
```

Comment:
funnily enough here you actually use the block :)

Reply:
Oh, yeah, good catch. This is actually wrong! We don't care about the latest known protocol version of the block producer, we want to know the epoch protocol version.